src/protocol/sort/generate_permutation.rs

```diff
@@ -126,38 +97,58 @@ pub(super) async fn malicious_shuffle_and_reveal_permutation<F: Field + Extendab
-    let sort_permutation = generate_permutation_opt(ctx.narrow(&SortKeys), sort_keys).await?;
-    shuffle_and_reveal_permutation(ctx.narrow(&ShuffleRevealPermutation), sort_permutation).await
-}
-
-/// This function takes in a semihonest context and sort keys, generates a sort permutation, shuffles and reveals it and
-/// returns both shuffle-revealed permutation and 2/3 randoms which were used to shuffle the permutation
-/// The output of this can be applied to any of semihonest/malicious context
-/// # Panics
-/// If unable to convert sort keys length to u32
-/// # Errors
-/// If unable to convert sort keys length to u32
-pub async fn malicious_generate_permutation_and_reveal_shuffled<F: Field + ExtendableField>(
-    sh_ctx: SemiHonestContext<'_>,
+pub async fn generate_permutation_and_reveal_shuffled<C, S, F>(
+    sh_ctx: C,
     sort_keys: impl Iterator<Item = &Vec<Vec<Replicated<F>>>>,
-) -> Result<RevealedAndRandomPermutations, Error> {
-    let (malicious_validator, sort_permutation) =
-        malicious_generate_permutation_opt(sh_ctx.narrow(&SortKeys), sort_keys).await?;
-
-    let m_ctx = malicious_validator.context();
-
-    malicious_shuffle_and_reveal_permutation(
+) -> Result<RevealedAndRandomPermutations, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + BasicProtocols<C::UpgradedContext<F>, F> + 'static,
+    F: ExtendableField,
+    ShuffledPermutationWrapper<S, C::UpgradedContext<F>>: DowngradeMalicious<Target = Vec<u32>>,
+{
+    let (validator, sort_permutation) =
+        generate_permutation_opt(sh_ctx.narrow(&SortKeys), sort_keys).await?;
+
+    let m_ctx = validator.context();
+    shuffle_and_reveal_permutation::<C, _, _>(
         m_ctx.narrow(&ShuffleRevealPermutation),
         sort_permutation,
-        malicious_validator,
+        validator,
     )
     .await
 }
 
+#[async_trait]
+impl<'a, F: ExtendableField> DowngradeMalicious
+    for ShuffledPermutationWrapper<MaliciousReplicated<F>, UpgradedMaliciousContext<'a, F>>
+{
+    type Target = Vec<u32>;
+    /// For ShuffledPermutationWrapper on downgrading, we reveal the permutation. This runs reveal on the malicious context
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+        let output = self
+            .reveal(self.ctx.narrow(&RevealPermutation), NoRecord)
+            .await
+            .unwrap();
+        UnauthorizedDowngradeWrapper::new(output)
+    }
+}
+
+#[async_trait]
+impl<'a, F: ExtendableField> DowngradeMalicious
+    for ShuffledPermutationWrapper<Replicated<F>, UpgradedSemiHonestContext<'a, F>>
+{
+    type Target = Vec<u32>;
+    async fn downgrade(self) -> UnauthorizedDowngradeWrapper<Self::Target> {
+        let output = self
+            .reveal(self.ctx.narrow(&RevealPermutation), NoRecord)
+            .await
+            .unwrap();
+        UnauthorizedDowngradeWrapper::new(output)
```

Comment:
I am a bit surprised that we need to downgrade via wrapper. Isn't it just semi-honest to semi-honest conversion?

Reply:
Oh, I meant to follow up on this one and it got lost in the noise.  @richajaindce I think that this is your work.  Is this because we need to reveal and the way we decided to reveal was to do that as part of the downgrade?

Reply:
This seems like something we can iterate on, for sure.  It seems like the biggest question is how we manage reveal.

What we might want to do is only implement Reveal for the two upgraded contexts.  That might be simpler overall.